src/connectors/exchangeRate/index.ts

```diff
@@ -0,0 +1,251 @@
+      from: data.from,
+      to: data.to,
+      rate: data.rate,
+      updatedAt: new Date(data.updatedAt),
+    }
+  }
+
+  private updateAllRatesAndGetRate = async ({
+    from,
+    to,
+  }: Pair): Promise<GQLExchangeRate | never> => {
+    logger.warn(
+      'exchangeRate requested APIs to get rates instead of from cache'
+    )
+    const tokenRates = await this.fetchTokenRates()
+    const fiatRates = await this.fetchFiatRates()
+    await this.updateRatesToCache(tokenRates)
+    await this.updateRatesToCache(fiatRates)
+    for (const rate of tokenRates) {
+      if (rate.from === from && rate.to === to) {
+        return rate
+      }
+    }
+    for (const rate of fiatRates) {
+      if (rate.from === from && rate.to === to) {
+        return rate
+      }
+    }
+    throw new UnknownError('Unexpected missing return value')
+  }
+
+  private updateRatesToCache = async (rates: GQLExchangeRate[]) => {
+    for (const rate of rates) {
+      this.cache.storeObject({
+        keys: this.genCacheKeys(rate),
+        data: rate,
+        expire: this.expire,
+      })
+    }
+  }
+
+  private genCacheKeys = (pair: Pair) => ({ id: pair.from + pair.to })
+
+  private getTokenPairs = () => {
+    const pairs = []
+    for (const t of tokenCurrencies) {
+      for (const q of quoteCurrencies) {
+        pairs.push({ from: t, to: q })
+      }
+    }
+    return pairs
+  }
+
+  private getFiatPairs = () => {
+    const pairs = []
+    for (const f of fiatCurrencies) {
+      for (const q of quoteCurrencies) {
+        pairs.push({ from: f, to: q })
+      }
+    }
+    return pairs
+  }
+
+  private fetchTokenRates = async (): Promise<GQLExchangeRate[]> => {
+    const data = await this.requestCoingeckoAPI(
+      tokenCurrencies,
+      quoteCurrencies
+    )
+    const rates: GQLExchangeRate[] = []
+    for (const t of tokenCurrencies) {
+      for (const q of quoteCurrencies) {
+        rates.push({
+          from: t,
+          to: q,
+          rate: data[TOKEN_TO_COINGECKO_ID[t]][q.toLowerCase()],
+          updatedAt: new Date(
+            data[TOKEN_TO_COINGECKO_ID[t]].last_updated_at * 1000
+          ),
+        })
+      }
+    }
+    return rates
+  }
+
+  private fetchFiatRates = async (): Promise<GQLExchangeRate[]> => {
+    const rates: GQLExchangeRate[] = []
+    for (const t of fiatCurrencies) {
+      const data = await this.requestExchangeRatesDataAPI(t, quoteCurrencies)
+      for (const q of quoteCurrencies) {
+        rates.push({
+          from: t,
+          to: q,
+          rate: data.rates[q],
+          updatedAt: new Date(data.timestamp * 1000),
+        })
+      }
+    }
+    return rates
+  }
+
+  private requestCoingeckoAPI = async (
+    bases: TokenCurrency[],
+    quotes: GQLQuoteCurrency[]
+  ): Promise<any | never> => {
+    const ids = bases.map((i) => TOKEN_TO_COINGECKO_ID[i]).join()
+    const vs_currencies = quotes.join()
+    try {
+      const reps = await axios.get(COINGECKO_API_URL, {
```

Comment:
There is no API key required from CoinGecko?

Reply:
Yes
> Our Free API* has a rate limit of 10-50 calls/minute, and doesn't require API key
see https://www.coingecko.com/en/api/documentation